include/dali/core/mm/pool_resource.h

```diff
@@ -139,18 +151,24 @@ class pool_resource_base : public memory_resource<Kind> {
     alignment = std::max(alignment, options_.upstream_alignment);
     size_t blk_size = bytes;
 
-    void *new_block = get_upstream_block(blk_size, bytes, alignment);
-    assert(new_block);
-    if (blk_size == bytes) {
-      // we've allocated a block exactly of the required size - there's little
-      // chance that it will be merged with anything in the pool, so we'll return it as-is
-      return new_block;
-    } else {
-      // we've allocated an oversized block - put the remainder in the free list
+    char *block_start = static_cast<char*>(get_upstream_block(blk_size, bytes, alignment));
+    assert(block_start);
+
+    char *ret = detail::align_ptr(block_start, alignment);
+    char *tail = ret + bytes;
+    char *block_end = block_start + blk_size;
+    assert(tail <= block_end);
+
+    if (blk_size != bytes) {
```

Comment:
```suggestion
    if (blk_size > bytes) {
```
According to what the comment says?

Reply:
Actually, it's already indirectly tested as `assert(tail <= block_end);`